File: csharp/Tdcv2.Tests/FileGenTest.cs
using Tdcv2.Engine;
using Tdcv2.Generators;
using Tdcv2.Parser;

namespace Tdcv2.Tests;

/// <summary>
/// The file generator, which the shared cases cannot reach because they carry no files.
/// </summary>
/// <remarks>
/// What matters here is not that a CSV parses — it is the handful of decisions that turn a plausible
/// reading of a file into a wrong one: a weight column that silently drops a product, a header row
/// counted as data, a byte-order mark hiding the first column's name.
/// </remarks>
public class FileGenTest : IDisposable
{
    private readonly string _dir = Directory.CreateDirectory(
        Path.Combine(Path.GetTempPath(), "tdcv2-filegen-" + Guid.NewGuid().ToString("N"))).FullName;

    public void Dispose() => Directory.Delete(_dir, recursive: true);

    private string Write(string name, string content)
    {
        string path = Path.Combine(_dir, name);
        File.WriteAllText(path, content);
        return path;
    }

    [Fact]
    public void APlainListIsOneValuePerLineWithBlanksSkipped()
    {
        Write("codes.txt", "alpha\n\n  beta  \ngamma\n");
        var attrs = new Dictionary<string, string> { ["src"] = "codes.txt" };

        Assert.Equal(new[] { "alpha", "beta", "gamma" }, FileGen.Load(attrs, _dir));
    }

    [Fact]
    public void ANamedColumnSkipsTheHeaderAndANumberedOneDoesNot()
    {
        Write("rows.csv", "name,count\nSmith,3\nJones,1\n");

        Assert.Equal(
            new[] { "Smith", "Jones" },
            FileGen.Load(new Dictionary<string, string> { ["src"] = "rows.csv", ["column"] = "name" }, _dir));

        // A file of pure data has no header to skip, so a numbered column takes every row.
        Assert.Equal(
            new[] { "name", "Smith", "Jones" },
            FileGen.Load(new Dictionary<string, string> { ["src"] = "rows.csv", ["column"] = "1" }, _dir));
    }

    [Fact]
    public void AByteOrderMarkDoesNotHideTheFirstColumnsName()
    {
        // Excel writes one ahead of the first header cell; without stripping it, every
        // "Save as CSV" export would fail to resolve its first column by name and no other.
        Write("bom.csv", "﻿name,count\nSmith,3\n");

        Assert.Equal(
            new[] { "Smith" },
            FileGen.Load(new Dictionary<string, string> { ["src"] = "bom.csv", ["column"] = "name" }, _dir));
    }

    [Fact]
    public void QuotedFieldsKeepTheirCommasAndTheirDoubledQuotes()
    {
        Write("quoted.csv", "name\n\"Smith, John\"\n\"He said \"\"hi\"\"\"\n");

        Assert.Equal(
            new[] { "Smith, John", "He said \"hi\"" },
            FileGen.Load(new Dictionary<string, string> { ["src"] = "quoted.csv", ["column"] = "name" }, _dir));
    }

    [Fact]
    public void WeightsAreHonouredExactlyRatherThanOnAverage()
    {
        Write("weighted.csv", "name,count\nSmith,20000\nJones,10000\n");
        var attrs = new Dictionary<string, string>
        {
            ["src"] = "weighted.csv", ["column"] = "name", ["weight"] = "count",
        };

        FileGen.Weighted weighted = FileGen.LoadWeighted(attrs, _dir)!;
        Assert.Equal(new[] { "Smith", "Jones" }, weighted.Values);
        Assert.Equal(200.0 / 3, weighted.Percents[0], 9);
        Assert.Equal(100.0 / 3, weighted.Percents[1], 9);
    }

    [Fact]
    public void AZeroWeightDropsAValueButABlankOneIsAnError()
    {
        Write("zero.csv", "name,count\nSmith,5\nGone,0\n");
        var attrs = new Dictionary<string, string>
        {
            ["src"] = "zero.csv", ["column"] = "name", ["weight"] = "count",
        };
        Assert.Equal(new[] { "Smith" }, FileGen.LoadWeighted(attrs, _dir)!.Values);

        // A blank must not slide through as zero. A product vanishing from a catalogue because one
        // cell of an export was empty is discovered far too late, and "missing" and "deliberately
        // excluded" are different statements — only one of them is actionable.
        Write("blank.csv", "name,count\nSmith,5\nOops,\n");
        var blank = new Dictionary<string, string>
        {
            ["src"] = "blank.csv", ["column"] = "name", ["weight"] = "count",
        };
        ArgumentException e = Assert.Throws<ArgumentException>(() => FileGen.LoadWeighted(blank, _dir));
        Assert.Contains("write 0 to exclude it", e.Message);
    }

    [Fact]
    public void TwoColumnsOnOneRowLinkComeFromTheSameRecord()
    {
        Write("places.csv", "city,postcode\nBerlin,10115\nHamburg,20095\nMunich,80331\n");

        string rendered = Engines.Render(
            ConfigBuilder.Build(
                TdcParserFacade.Parse(
                    "<tdc><env count=\"4\" seed=\"link\" mode=\"memory\">"
                    + "<sequence name=\"City\"><gen type=\"file\" src=\"places.csv\" "
                    + "column=\"city\" row=\"place\"/></sequence>"
                    + "<sequence name=\"Zip\"><gen type=\"file\" src=\"places.csv\" "
                    + "column=\"postcode\" row=\"place\"/></sequence>"
                    + "</env><block><line><data>${{City}} ${{Zip}}</data></line></block></tdc>").Tree),
            null, null, _dir);

        // Every emitted pair is a real record, never a plausible-looking recombination.
        var real = new Dictionary<string, string>
        {
            ["Berlin"] = "10115", ["Hamburg"] = "20095", ["Munich"] = "80331",
        };
        foreach (string line in rendered.TrimEnd('\n').Split('\n'))
        {
            string[] parts = line.Split(' ');
            Assert.Equal(real[parts[0]], parts[1]);
        }
    }

    [Fact]
    public void ARelativeSourceIsRelativeToTheConfigNotTheWorkingDirectory()
    {
        string path = Write("here.txt", "one\n");

        // Otherwise the same config would work from one shell and fail from another.
        Assert.Equal(path, FileGen.Resolve("here.txt", _dir, null));
        Assert.Equal(path, FileGen.Resolve("./here.txt", _dir, null));
    }

    [Fact]
    public void TheDataAliasNamesTheConfiguredFoldersAndSaysSoWhenThereAreNone()
    {
        string path = Write("shared.txt", "one\n");
        Assert.Equal(path, FileGen.Resolve("@data/shared.txt", "/nowhere", new[] { _dir }));

        ArgumentException e = Assert.Throws<ArgumentException>(
            () => FileGen.Resolve("@data/shared.txt", _dir, null));
        Assert.Contains("needs at least one data folder", e.Message);
    }

    [Theory]
    [InlineData("tab")]
    [InlineData("\\t")]
    // A real tab arrives as a single character and is taken as written: trimming it would leave
    // nothing and fall back to a comma, so resolving twice has to be harmless.
    [InlineData("\t")]
    public void EverySpellingOfTabIsATab(string spec)
    {
        Write("tabbed.tsv", "name\tcount\nSmith\t3\n");
        var attrs = new Dictionary<string, string>
        {
            ["src"] = "tabbed.tsv", ["column"] = "name", ["delimiter"] = spec,
        };

        Assert.Equal(new[] { "Smith" }, FileGen.Load(attrs, _dir));
    }

    [Fact]
    public void SequentialOrderWalksTheFileInOrder()
    {
        Write("months.txt", "Jan\nFeb\nMar\n");

        string rendered = Engines.Render(
            ConfigBuilder.Build(
                TdcParserFacade.Parse(
                    "<tdc><env count=\"5\" seed=\"s\" mode=\"memory\"><sequence name=\"M\">"
                    + "<gen type=\"file\" src=\"months.txt\" order=\"sequential\"/></sequence>"
                    + "</env><block><line><data>${{M}}</data></line></block></tdc>").Tree),
            null, null, _dir);

        // Looping is the default: a short list over many rows is the ordinary case.
        Assert.Equal("Jan\nFeb\nMar\nJan\nFeb\n", rendered);
    }
}

File: java/src/test/java/io/github/nickliapin/tdc/FileGenTest.java
package io.github.nickliapin.tdc;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.HashSet;
import java.util.Set;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

/** {@code <gen type="file">} — the generator that reads the user's own data. */
class FileGenTest {

  private static TDC tdc(Path dir, String body) {
    return TDC.options()
        .configString(
            "<tdc><env mode=\"memory\" count=\"20\" seed=\"file\" local=\"en\">"
                + body
                + "</env><block><line><data>${{V}}</data></line></block></tdc>")
        .baseDir(dir)
        .build();
  }

  private static Set<String> valuesOf(TDC tdc) {
    Set<String> seen = new HashSet<>();
    for (TDC.Row row : tdc.iterate()) {
      seen.add(row.get("V"));
    }
    return seen;
  }

  @Test
  @DisplayName("a plain list file, one value per line")
  void listFile(@TempDir Path dir) throws IOException {
    Files.writeString(dir.resolve("codes.txt"), "alpha\n\n  beta  \ngamma\n");
    Set<String> seen = valuesOf(tdc(dir, "<sequence name=\"V\"><gen type=\"file\" src=\"codes.txt\"/></sequence>"));
    // Blank lines are skipped and values are trimmed, so " beta " is "beta".
    assertEquals(Set.of("alpha", "beta", "gamma"), seen);
  }

  @Test
  @DisplayName("a CSV column by header name")
  void csvByName(@TempDir Path dir) throws IOException {
    Files.writeString(dir.resolve("users.csv"), "id,email,city\n1,a@x.io,Perm\n2,b@x.io,Omsk\n");
    Set<String> seen =
        valuesOf(tdc(dir, "<sequence name=\"V\"><gen type=\"file\" src=\"users.csv\" column=\"email\"/></sequence>"));
    assertEquals(Set.of("a@x.io", "b@x.io"), seen);
  }

  @Test
  @DisplayName("a CSV column by 1-based position keeps the first row unless told otherwise")
  void csvByPosition(@TempDir Path dir) throws IOException {
    Files.writeString(dir.resolve("data.csv"), "one,two\nthree,four\n");
    // A numbered column means the file may have no header, so nothing is skipped by default.
    assertEquals(
        Set.of("one", "three"),
        valuesOf(tdc(dir, "<sequence name=\"V\"><gen type=\"file\" src=\"data.csv\" column=\"1\"/></sequence>")));
    assertEquals(
        Set.of("three"),
        valuesOf(
            tdc(
                dir,
                "<sequence name=\"V\"><gen type=\"file\" src=\"data.csv\" column=\"1\" header=\"true\"/></sequence>")));
  }

  @Test
  @DisplayName("quoted fields survive the delimiter and the quotes inside them")
  void quotedFields(@TempDir Path dir) throws IOException {
    Files.writeString(
        dir.resolve("products.csv"),
        "name,price\n\"Knife set, 3 pcs\",10\n\"say \"\"hi\"\"\",20\n");
    assertEquals(
        Set.of("Knife set, 3 pcs", "say \"hi\""),
        valuesOf(
            tdc(dir, "<sequence name=\"V\"><gen type=\"file\" src=\"products.csv\" column=\"name\"/></sequence>")));
  }

  @Test
  @DisplayName("delimiters can be named, and a real tab is taken as written")
  void delimiters(@TempDir Path dir) throws IOException {
    Files.writeString(dir.resolve("t.csv"), "a\tb\nx\ty\n");
    assertEquals(
        Set.of("x"),
        valuesOf(
            tdc(dir, "<sequence name=\"V\"><gen type=\"file\" src=\"t.csv\" column=\"a\" delimiter=\"tab\"/></sequence>")));
    Files.writeString(dir.resolve("s.csv"), "a;b\np;q\n");
    assertEquals(
        Set.of("p"),
        valuesOf(
            tdc(dir, "<sequence name=\"V\"><gen type=\"file\" src=\"s.csv\" column=\"a\" delimiter=\";\"/></sequence>")));
  }

  @Test
  @DisplayName("a byte-order mark does not hide the first column's name")
  void byteOrderMark(@TempDir Path dir) throws IOException {
    // Excel writes one ahead of the first header cell on every "Save as CSV".
    Files.writeString(dir.resolve("excel.csv"), "﻿id,name\n7,Ann\n");
    assertEquals(
        Set.of("7"),
        valuesOf(tdc(dir, "<sequence name=\"V\"><gen type=\"file\" src=\"excel.csv\" column=\"id\"/></sequence>")));
  }

  @Test
  @DisplayName("a relative src resolves against the config file, not the working directory")
  void relativeToTheConfig(@TempDir Path dir) throws IOException {
    Path sub = Files.createDirectories(dir.resolve("configs"));
    Files.writeString(sub.resolve("codes.txt"), "only\n");
    Files.writeString(
        sub.resolve("run.tdc"),
        """
        <tdc><env mode="memory" count="2" seed="rel" local="en">
          <sequence name="V"><gen type="file" src="codes.txt"/></sequence>
        </env><block><line><data>${{V}}</data></line></block></tdc>
        """);
    assertEquals("only\nonly\n", new TDC(sub.resolve("run.tdc")).toString());
  }

  @Test
  @DisplayName("a missing file, an empty one, and an unknown column are all reported")
  void failures(@TempDir Path dir) throws IOException {
    assertThrows(
        RuntimeException.class,
        () -> tdc(dir, "<sequence name=\"V\"><gen type=\"file\" src=\"nope.txt\"/></sequence>").toString());

    Files.writeString(dir.resolve("empty.txt"), "\n\n");
    assertThrows(
        IllegalArgumentException.class,
        () -> tdc(dir, "<sequence name=\"V\"><gen type=\"file\" src=\"empty.txt\"/></sequence>").toString());

    Files.writeString(dir.resolve("u.csv"), "a,b\n1,2\n");
    IllegalArgumentException e =
        assertThrows(
            IllegalArgumentException.class,
            () ->
                tdc(dir, "<sequence name=\"V\"><gen type=\"file\" src=\"u.csv\" column=\"missing\"/></sequence>")
                    .toString());
    assertTrue(e.getMessage().contains("was not found in the header row"), e.getMessage());
  }
}
